File: apps/api/mosaera_api/routes/sessions.py
"""PM chat session routes: list / create / rename / archive a project's conversation threads.

A session scopes chat HISTORY (issue #30). Project knowledge (brief/backlog/runs/context
registry) stays project-scoped and shared across sessions — so switching sessions changes the
conversation, not what Quincy knows about the project. Message read/write themselves live in
``routes/messages.py`` (session-scoped via ``?session_id=`` / the message body)."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException

from mosaera_api.routes.context import AppContext
from mosaera_api.schemas import SessionCreateBody, SessionPatchBody


def make_sessions_router(ctx: AppContext) -> APIRouter:
    api = APIRouter()

    def _require_project(mem: Any, project_id: str) -> None:
        if mem.project_detail(project_id) is None:
            raise HTTPException(status_code=404, detail="unknown project")

    def _require_session(mem: Any, project_id: str, session_id: str) -> dict[str, Any]:
        # A session is always addressed under its project — a mismatched pair 404s so one
        # project can never read or mutate another's threads.
        sess = mem.get_pm_session(session_id)
        if sess is None or sess["project_id"] != project_id:
            raise HTTPException(status_code=404, detail="unknown session")
        return sess

    @api.get("/projects/{project_id}/sessions")
    def list_sessions(project_id: str, include_archived: bool = False) -> dict[str, Any]:
        mem = ctx.require_memory()
        _require_project(mem, project_id)
        return {
            "sessions": mem.list_pm_sessions(project_id, include_archived=include_archived),
        }

    @api.post("/projects/{project_id}/sessions", status_code=201)
    def create_session(project_id: str, body: SessionCreateBody) -> dict[str, Any]:
        mem = ctx.require_memory()
        _require_project(mem, project_id)
        session_id = mem.create_pm_session(project_id, title=body.title.strip()[:256])
        created = mem.get_pm_session(session_id)
        if created is None:  # pragma: no cover — just-created row must exist
            raise HTTPException(status_code=500, detail="session create failed")
        return created

    @api.patch("/projects/{project_id}/sessions/{session_id}")
    def patch_session(project_id: str, session_id: str, body: SessionPatchBody) -> dict[str, Any]:
        mem = ctx.require_memory()
        _require_project(mem, project_id)
        _require_session(mem, project_id, session_id)
        if body.title is not None:
            mem.rename_pm_session(session_id, body.title.strip()[:256])
        if body.archived is not None:
            mem.set_pm_session_archived(session_id, body.archived)
        updated = mem.get_pm_session(session_id)
        if updated is None:  # pragma: no cover — existence checked above
            raise HTTPException(status_code=404, detail="unknown session")
        return updated

    return api

File: apps/web/src/api/sessions.ts
// PM chat sessions (issue #30): per-project conversation threads. Kept in its own module
// (not client.ts, which is grandfathered over the size ceiling) — the message read/write
// helpers stay in client.ts; this owns the session lifecycle surface.

import { apiFetch } from "./auth";

/** One PM conversation thread within a project. History is scoped to a session; project
 *  knowledge (brief/backlog/runs/context) is shared across a project's sessions. */
export interface PmSession {
  id: string;
  project_id: string;
  title: string; // "" until the first user turn auto-names it
  created_at: string | null;
  updated_at: string | null;
  message_count: number;
}

async function json<T>(res: Response): Promise<T> {
  if (!res.ok) {
    const detail = await res.text().catch(() => "");
    throw new Error(`${res.status} ${res.statusText}${detail ? `: ${detail}` : ""}`);
  }
  return res.json() as Promise<T>;
}

export const sessionsApi = {
  list: (projectId: string) =>
    apiFetch(
      `/api/projects/${projectId}/sessions`,
    ).then(json<{ sessions: PmSession[] }>),

  create: (projectId: string, title = "") =>
    apiFetch(`/api/projects/${projectId}/sessions`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ title }),
    }).then(json<PmSession>),

  patch: (projectId: string, sessionId: string, patch: { title?: string; archived?: boolean }) =>
    apiFetch(`/api/projects/${projectId}/sessions/${sessionId}`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(patch),
    }).then(json<PmSession>),
};
